### PR TITLE
Add Polymer2MixinScanner

### PR DESCRIPTION
**Pull request by justinfagnani**
- [x] CHANGELOG.md has been updated

### CODE REVIEW

File: src/model/element-mixin.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {Element, ScannedElement} from './element';
+import {Document} from './model';
+
+export {Attribute} from './element';
+
+export class ScannedElementMixin extends ScannedElement {
+  get name() {
```
Comment (rictic) [inline src/model/element-mixin.ts]: What's with this name property?

File: src/javascript/jsdoc.ts [hunk 2]
```diff
@@ -74,6 +74,11 @@ function parsePolymerElement(tag: doctrine.Tag): Tag {
   return {tag: tag.title, type: null, name: tag.description, description: null};
 }
 
+// @polymerMixin [name]
+function parsePolymerMixin(tag: doctrine.Tag): Tag {
+  return {tag: tag.title, type: null, name: tag.description, description: null};
```
Comment (rictic) [inline src/javascript/jsdoc.ts]: How does this behave with annotations that include a description, like

```javascript
/**
 * @polymerMixin MyMixin A mixin that adds random yodeling through the web audio apis.
 */
const MyMixin = function(clazz) {...};
```

or even:

```javascript
/**
 * @polymerMixin A mixin that adds random yodeling through the web audio apis.
 */
const MyMixin = function(clazz) {...};
```

File: src/polymer/polymer-element.ts [hunk 3]
```diff
@@ -67,7 +67,7 @@ export interface Options {
   events?: ScannedEvent[];
 
   abstract?: boolean;
-  sourceRange: SourceRange|undefined;
+  sourceRange?: SourceRange|undefined;
```
Comment (rictic) [inline src/polymer/polymer-element.ts]: Why loosen this?

File: src/polymer/polymer2-config.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
+import * as estree from 'estree';
```
Comment (rictic) [inline src/polymer/polymer2-config.ts]: Need license

File: src/polymer/polymer2-config.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+import {getIdentifierName} from '../javascript/ast-value';
+import {getPropertyValue} from '../javascript/esutil';
+import {JavaScriptDocument} from '../javascript/javascript-document';
+
+import {analyzeProperties} from './analyze-properties';
+import {ScannedPolymerProperty} from './polymer-element';
+
+export function getConfig(node: estree.ClassDeclaration|estree.ClassExpression):
```
Comment (rictic) [inline src/polymer/polymer2-config.ts]: Document this function. What's a config in this context?

Comment (rictic) [inline src/polymer/polymer2-config.ts]: Document this function. What's a config in this context?

File: src/polymer/polymer2-config.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  const configBody = config.value.body;
+  if (configBody.body.length !== 1) {
+    // not a single statement function
+    return null;
+  }
+  if (configBody.body[0].type !== 'ReturnStatement') {
+    return null;
+  }
+
+  const returnStatement = configBody.body[0] as estree.ReturnStatement;
```
Comment (rictic) [inline src/polymer/polymer2-config.ts]: Here and below with returnValue, if you do the assignment above the if statement then you don't need the type cast.

File: src/polymer/polymer2-mixin-scanner.ts [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    const docs = jsdoc.parseJsdoc(comment);
+    if (this._hasPolymerMixinDocTag(docs)) {
+      const name = node.id.name;
+      this._currentMixinFunction = node;
+      this._currentMixin = new ScannedPolymerElementMixin({
+        tagName: name,
+      });
+      this._currentMixinNode = node;
+      this._currentMixin.tagName = name;
+      this._mixins.push(this._currentMixin);
```
Comment (rictic) [inline src/polymer/polymer2-mixin-scanner.ts]: Here and elsewhere, we should give the best source range that we can for the mixin. When we add the mixin to the array, it seems like a good time to check to see if it has a source range yet, and if not, give it the best one available.

File: src/polymer/polymer2-mixin-scanner.ts [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+  }
+
+  enterVariableDeclaration(
+      node: estree.VariableDeclaration, _parent: estree.Node) {
+    const comment = esutil.getAttachedComment(node) || '';
+    const docs = jsdoc.parseJsdoc(comment);
+    const isMixin = this._hasPolymerMixinDocTag(docs);
+    if (isMixin) {
+      this._currentMixin = new ScannedPolymerElementMixin();
+      this._currentMixinNode = node;
```
Comment (rictic) [inline src/polymer/polymer2-mixin-scanner.ts]: Or maybe the thing to do is to set the source range here when you initially create the mixin, and then if we encounter something better in our walk later it'll overwrite it?

File: src/test/polymer/polymer2-mixin-scanner_test.ts [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+import * as fs from 'fs';
+import * as path from 'path';
+
+import {Visitor} from '../../javascript/estree-visitor';
+import {JavaScriptParser} from '../../javascript/javascript-parser';
+import {ScannedFeature} from '../../model/model';
+import {ScannedPolymerElementMixin} from '../../polymer/polymer-element-mixin';
+import {Polymer2MixinScanner} from '../../polymer/polymer2-mixin-scanner';
+
+// function compareNames(a: {name?: string}, b: {name?: string}): number {
```
Comment (rictic) [inline src/test/polymer/polymer2-mixin-scanner_test.ts]: delete?

File: src/polymer/polymer2-mixin-scanner.ts [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+  enterFunctionDeclaration(
+      node: estree.FunctionDeclaration, _parent: estree.Node) {
+    const comment = esutil.getAttachedComment(node) || '';
+    const docs = jsdoc.parseJsdoc(comment);
+    if (this._hasPolymerMixinDocTag(docs)) {
+      const name = node.id.name;
+      this._currentMixinFunction = node;
+      this._currentMixin = new ScannedPolymerElementMixin({
+        tagName: name,
```
Comment (rictic) [inline src/polymer/polymer2-mixin-scanner.ts]: I don't think we should overload tagName like this. Wouldn't it be better to just have a name property on ScannedPolymerElementMixin?

File: src/polymer/polymer2-mixin-scanner.ts [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+  async scan(
+      document: JavaScriptDocument, visit: (visitor: Visitor) => Promise<void>):
+      Promise<ScannedPolymerElementMixin[]> {
+    let visitor = new MixinVisitor(document);
+    await visit(visitor);
+    return visitor._mixins;
+  }
+}
+
+class MixinVisitor implements Visitor {
```
Comment (rictic) [inline src/polymer/polymer2-mixin-scanner.ts]: I think we should be generous with warnings here for mixins that are explicitly declared but that we can't analyze. We could have an additional annotation that indicates that the analyzer won't be able to figure out a more complex mixin, and that it's the user's responsibility at that point to explain it to us.

If we don't, we're in danger of being opaque, mysterious and thus frustrating to use.

File: test/mocha.opts [hunk 12]
```diff
@@ -1,3 +1,2 @@
 --ui tdd
 --require source-map-support/register
-lib/test/*_test.js lib/test/**/*_test.js
```
Comment (rictic) [inline test/mocha.opts]: revert?

Comment (rictic) [inline test/mocha.opts]: Or, hm, rather I think this should be put into the package.json

Though I've had good luck lately with using `.only`, it seems mocha fixed the bugs I was seeing with it at some point.

File: src/test/polymer/polymer2-mixin-scanner_test.ts [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+//   const nameB = b && b.name;
+
+//   if (nameA == null)
+//     return (nameB == null) ? 0 : -1;
+//   if (nameB == null)
+//     return 1;
+//   return nameA.localeCompare(nameB);
+// };
+
+suite('Polymer2MixinScanner', () => {
```
Comment (rictic) [inline src/test/polymer/polymer2-mixin-scanner_test.ts]: Please use the underline printer to also test source ranges.

**Comment by rictic** [comment]
Oh, and there's a couple of lint errors, fixable with `tslint --fix`

File: src/model/element-mixin.ts (hunk 1 above)
Comment (justinfagnani) [inline src/model/element-mixin.ts]: I wanted a better name than `tagName`. I'm going to pull out a common subclass.

File: src/javascript/jsdoc.ts (hunk 2 above)
Comment (rictic) [inline src/javascript/jsdoc.ts]: Please add a TODO to the code here, file a bug, or add a skipped test, just to help us keep track of it.

Ah, nm, found the TODO later on.

File: src/polymer/polymer-element.ts [hunk 14]
```diff
@@ -80,10 +81,61 @@ export interface Options {
   sourceRange: SourceRange|undefined;
 }
 
+export interface ScannedPolymerExtension extends ScannedElementBase {
```
Comment (rictic) [inline src/polymer/polymer-element.ts]: Arg, this is painful. Yeah, I can definitely see why we want mixins here.

Add a TODO to use native mixins ourselves here once we drop node 4?

File: src/polymer/polymer2-config.ts [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+import {getIdentifierName} from '../javascript/ast-value';
+import {getPropertyValue} from '../javascript/esutil';
+import {JavaScriptDocument} from '../javascript/javascript-document';
+
+import {analyzeProperties} from './analyze-properties';
+import {ScannedPolymerProperty} from './polymer-element';
+
+/**
+ * Returns the object literal that defines a Polymer element configuration from
+ * a Plymer element class.
```
Comment (rictic) [inline src/polymer/polymer2-config.ts]: Plymer -> Polymer

Comment (justinfagnani) [inline src/polymer/polymer2-config.ts]: fixed in next PR

**Pull request merged**